// ng-otp/bundles/ng-otp.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('rxjs/operators'), require('@angular/core'), require('@angular/forms'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('ng-otp', ['exports', 'rxjs', 'rxjs/operators', '@angular/core', '@angular/forms', '@angular/platform-browser'], factory) :
    (factory((global['ng-otp'] = {}),global.rxjs,global.rxjs.operators,global.ng.core,global.ng.forms,global.ng.platformBrowser));
}(this, (function (exports,rxjs,operators,i0,forms,platformBrowser) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgOtpService = /** @class */ (function () {
        function NgOtpService() {
        }
        /**
         * @param {?} string
         * @return {?}
         */
        NgOtpService.prototype.isEmptySting = /**
         * @param {?} string
         * @return {?}
         */
            function (string) {
                return string === '';
            };
        /**
         * @param {?} index
         * @param {?} limit
         * @return {?}
         */
        NgOtpService.prototype.isLastInput = /**
         * @param {?} index
         * @param {?} limit
         * @return {?}
         */
            function (index, limit) {
                return index === limit;
            };
        /**
         * @param {?} index
         * @return {?}
         */
        NgOtpService.prototype.getElement = /**
         * @param {?} index
         * @return {?}
         */
            function (index) {
                return document.querySelector("#otp-" + index);
            };
        NgOtpService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        NgOtpService.ctorParameters = function () { return []; };
        /** @nocollapse */ NgOtpService.ngInjectableDef = i0.defineInjectable({ factory: function NgOtpService_Factory() { return new NgOtpService(); }, token: NgOtpService, providedIn: "root" });
        return NgOtpService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgOtpComponent = /** @class */ (function () {
        function NgOtpComponent(formBuilder, ngOtpService) {
            var _this = this;
            this.formBuilder = formBuilder;
            this.ngOtpService = ngOtpService;
            this._allowedCharacters = /./;
            this._typeOfInput = 'text';
            this.keyboardType = 'text';
            this.otpOut = new i0.EventEmitter();
            this.limitArray = [];
            this.isKeyAcceptable = true;
            this.changeFocus$ = new rxjs.Subject();
            this.subscription = new rxjs.Subscription();
            this.subscription.add(this.changeFocus$
                .pipe(operators.throttleTime(50)).subscribe(( /**
         * @param {?} index
         * @return {?}
         */function (index) { _this.changeFocus(index); })));
        }
        Object.defineProperty(NgOtpComponent.prototype, "allowedCharacters", {
            set: /**
             * @param {?} el
             * @return {?}
             */ function (el) {
                if (el) {
                    this._allowedCharacters = el;
                }
                else {
                    this.throwErrorForUndefinedElement(el);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgOtpComponent.prototype, "typeOfInput", {
            get: /**
             * @return {?}
             */ function () {
                return this._typeOfInput;
            },
            set: /**
             * @param {?} type
             * @return {?}
             */ function (type) {
                if (type) {
                    this._typeOfInput = type;
                    this.keyboardType = type === 'password' ? 'numeric' : 'text';
                }
                else {
                    this.throwErrorForUndefinedElement(type);
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NgOtpComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.limit = this.limit ? this.limit : 4;
                this.setFormBuilder();
            };
        /**
         * @return {?}
         */
        NgOtpComponent.prototype.setFormBuilder = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.otpForm = this.formBuilder.group({});
                this.limitArray = Array.from(Array(this.limit).keys());
                this.limitArray.map(( /**
                 * @param {?} element
                 * @return {?}
                 */function (element) {
                    _this.otpForm.addControl("otp-" + element, new forms.FormControl('', forms.Validators.required));
                }));
            };
        /**
         * @param {?} id
         * @return {?}
         */
        NgOtpComponent.prototype.changeFocus = /**
         * @param {?} id
         * @return {?}
         */
            function (id) {
                console.log(this.isKeyAcceptable);
                if (!this.isKeyAcceptable) {
                    this.isKeyAcceptable = true;
                    return;
                }
                /** @type {?} */
                var currentElement = this.ngOtpService.getElement(id);
                if (id && this.ngOtpService.isEmptySting(currentElement.value)) {
                    this.moveBackward(id);
                }
                else if (this.ngOtpService.isLastInput(id, this.limit - 1)) {
                    currentElement.select();
                    this.otpOut.emit(Object.values(this.otpForm.value).join(''));
                }
                else if (!this.ngOtpService.isEmptySting(currentElement.value)) {
                    this.moveForward(id);
                }
            };
        /**
         * @private
         * @param {?} id
         * @return {?}
         */
        NgOtpComponent.prototype.moveForward = /**
         * @private
         * @param {?} id
         * @return {?}
         */
            function (id) {
                /** @type {?} */
                var nextElement = this.ngOtpService.getElement(id + 1);
                nextElement.focus();
            };
        /**
         * @private
         * @param {?} id
         * @return {?}
         */
        NgOtpComponent.prototype.moveBackward = /**
         * @private
         * @param {?} id
         * @return {?}
         */
            function (id) {
                /** @type {?} */
                var nextElement = this.ngOtpService.getElement(id - 1);
                nextElement.focus();
            };
        /**
         * @param {?} id
         * @return {?}
         */
        NgOtpComponent.prototype.onFocus = /**
         * @param {?} id
         * @return {?}
         */
            function (id) {
                /** @type {?} */
                var currentElement = this.ngOtpService.getElement(id);
                currentElement.select();
            };
        /**
         * @return {?}
         */
        NgOtpComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.subscription.unsubscribe();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgOtpComponent.prototype.onKeyDown = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (event.key && (event.key !== 'Backspace' && event.key !== 'Delete')) {
                    if (this._allowedCharacters) {
                        if (this._allowedCharacters instanceof RegExp) {
                            if (!this._allowedCharacters.test(event.key)) {
                                this.isKeyAcceptable = false;
                                event.preventDefault();
                            }
                        }
                        else {
                            if (!this._allowedCharacters.includes(event.key)) {
                                this.isKeyAcceptable = false;
                                event.preventDefault();
                            }
                        }
                    }
                }
            };
        /**
         * @private
         * @template T
         * @param {?} element
         * @return {?}
         */
        NgOtpComponent.prototype.throwErrorForUndefinedElement = /**
         * @private
         * @template T
         * @param {?} element
         * @return {?}
         */
            function (element) {
                throw new Error("Is not possibile set " + element + " to this parameter");
            };
        NgOtpComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'ng-otp',
                        template: "<div\n  [formGroup]=\"otpForm\"\n  class=\"otp-wrap\"\n>\n  <input\n    [type]=\"typeOfInput\"\n    class=\"otp-input\"\n    maxlength=\"1\"\n    id=\"otp-{{i}}\"\n    *ngFor=\"let control of limitArray; let i = index\"\n    placeholder=\"-\"\n    (keyup)=\"changeFocus$.next(i)\"\n    (keydown)=\"onKeyDown($event)\"\n    (focusin)=\"onFocus(i)\"\n    formControlName=\"otp-{{i}}\"\n    [attr.inputmode]=\"keyboardType\"\n  >\n</div>",
                        styles: [".otp-input{width:50px;height:45px;background:0 0;border:none;text-align:center;font-size:25px;font-weight:600;outline:0}.otp-input::-moz-selection{background:#f0f9ff}.otp-input::selection{background:#f0f9ff}.otp-wrap{border-bottom:2px solid #d7d7d7;max-width:-webkit-max-content;max-width:-moz-max-content;max-width:max-content}"]
                    }] }
        ];
        /** @nocollapse */
        NgOtpComponent.ctorParameters = function () {
            return [
                { type: forms.FormBuilder },
                { type: NgOtpService }
            ];
        };
        NgOtpComponent.propDecorators = {
            limit: [{ type: i0.Input }],
            allowedCharacters: [{ type: i0.Input }],
            typeOfInput: [{ type: i0.Input }],
            otpOut: [{ type: i0.Output }]
        };
        return NgOtpComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgOtpModule = /** @class */ (function () {
        function NgOtpModule() {
        }
        NgOtpModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [NgOtpComponent],
                        imports: [
                            platformBrowser.BrowserModule,
                            forms.ReactiveFormsModule
                        ],
                        exports: [NgOtpComponent]
                    },] }
        ];
        return NgOtpModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NgOtpService = NgOtpService;
    exports.NgOtpComponent = NgOtpComponent;
    exports.NgOtpModule = NgOtpModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
